### PR TITLE
Tokenauth manas0410

### PR DESCRIPTION
**Pull request by Manas0410**
This PR covers following modification :

1.  UserToken Model for saving Tokens in DB.
2. Modified GenerateToken method to store tokens in db after their birth.
3. Added error handling for ValidateToken method .

**Followed Blacklisting Token technique for below modifications :**

4. Modified Regeneratetoken Method to regenerate the token after checking that token is valid and not destroyed.
5. Created DestroyToken method by blacklisting tokens from backend.

**Bug Fix**

6. Bug-fix in userSignUp cotroller , where Find method method used which returns empty array as a response and bcz of which it was always going inside error fallback which is inside below if condition 
 
```
const hasEmailExists = await UserModel.find({ email: req.body.email });
    if (hasEmailExists) {
      return res.send({
        data: {},
        meta: { message: "This email is already in use" },
      });
    }
```

As solution of above problem i have replaced find method with findOne which returns falsy value incase of no data found.

### CODE REVIEW

File: server/services/jwt/jwt.service.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    const refreshToken = jwt.sign(payload, JWT_PWD, {
+      expiresIn: REFRESH_EXP_IN,
+    });
+
+    // Find user token data in the database
+    const userTokenData = await userTokenModel.findOne({
+      userId: payload.userId,
+    });
+
+    if (userTokenData) {
```
Comment (xDAnkit) [inline server/services/jwt/jwt.service.js]: @Manas0410 - Can we use guard block approach here?

```
if(!userTokenData){
 // handle no token response
 return
}

// Rest code
```

File: server/services/jwt/jwt.service.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  // I(manas0410) am using token blacklisting
+
+  try {
+    const decoded = jwt.verify(refreshToken, JWT_PWD);
+
+    const userData = await userTokenModel.findOne({
+      userId: decoded.userId,
+    });
+
+    if (userData && userData?.refreshTokens.includes(refreshToken)) {
```
Comment (xDAnkit) [inline server/services/jwt/jwt.service.js]: @Manas0410 - Apply guard block approach

File: server/services/jwt/jwt.service.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+/**
+ * @description Method to destroy a JWT token
+ * @param {string} token - The JWT token to destroy
+ * @returns {Object} Confirmation message of token destruction
+ *  I(manas0410) am using token blacklisting
+ */
+export const destroyToken = async (token, type) => {
+  try {
+    if (type !== "refreshTokens" && type !== "accessTokens") {
```
Comment (xDAnkit) [inline server/services/jwt/jwt.service.js]:
```
if(!["accessTokens", "refreshTokens"].includes(type)){
 // Logic
}
```

File: server/services/jwt/jwt.service.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    );
+
+    // Check if the update operation was successful
+    if (result.nModified === 0) {
+      throw new Error("Token not found or user not authorized.");
+    }
+
+    return {
+      message:
+        "Token destruction should be handled on the client side or by maintaining a deny list on the server.",
```
Comment (xDAnkit) [inline server/services/jwt/jwt.service.js:131]: @Manas0410 - Can we move all hardcoded messages to a property or constant file

**Review by xDAnkit — CHANGES_REQUESTED** [review]
@Manas0410 - Pls check my comments

File: server/services/jwt/jwt.service.js (hunk 1 above)
Comment (Manas0410) [reply server/services/jwt/jwt.service.js]: Fixed

File: server/services/jwt/jwt.service.js (hunk 2 above)
Comment (Manas0410) [reply server/services/jwt/jwt.service.js]: Fixed

File: server/services/jwt/jwt.service.js (hunk 4 above)
Comment (Manas0410) [reply server/services/jwt/jwt.service.js:131]: @xDAnkit  - Fixed

File: server/services/jwt/jwt.service.js (hunk 3 above)
Comment (Manas0410) [reply server/services/jwt/jwt.service.js]: @xDAnkit  - Fixed

**Review by xDAnkit — APPROVED** [review]
Looks good to me

**Review by xDAnkit — APPROVED** [review]
Looks good to me

**Pull request merged**